### PR TITLE
Handle SonarLint compatible plugins

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

**Comment by dbmeneses** [comment]
LGTM

File: core/src/main/java/org/sonarsource/sonarlint/core/plugin/PluginCopier.java [hunk 1]
```diff
@@ -60,11 +63,18 @@ public PluginCopier(PluginCache fileCache, PluginIndexProvider pluginIndexProvid
     for (PluginReference pluginReference : pluginReferences) {
       File jarFile = getFromCacheOrCopy(pluginReference);
       PluginInfo info = PluginInfo.create(jarFile);
-      infosByKey.put(info.getKey(), info);
+      if ((info.isSonarLintSupported() != null && info.isSonarLintSupported().booleanValue()) || isWhitelisted(info.getKey())) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/plugin/PluginCopier.java]: BLOCKER NullPointerException might be thrown as 'isSonarLintSupported' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionInstaller.java [hunk 2]
```diff
@@ -34,24 +34,29 @@
...
       Plugin plugin = pluginRepository.getPluginInstance(pluginInfo.getKey());
-      Plugin.Context context = new Plugin.Context(sonarQubeVersion.get());
+      Plugin.Context context = new Plugin.Context(sqRuntime);
       plugin.define(context);
       for (Object extension : context.getExtensions()) {
-        if (!blacklisted(extension)) {
-          doInstall(container, matcher, pluginInfo, extension);
+        Boolean isSlPluginOrNull = pluginInfo.isSonarLintSupported();
+        if (isSlPluginOrNull != null && isSlPluginOrNull.booleanValue()) {
+          if (ExtensionUtils.isSonarLintSide(extension)) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionInstaller.java]: MAJOR Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionInstaller.java [hunk 3]
```diff
@@ -34,24 +34,29 @@
...
 
-  public ExtensionInstaller(SonarQubeVersion sonarQubeVersion, DefaultPluginRepository pluginRepository) {
-    this.sonarQubeVersion = sonarQubeVersion;
+  public ExtensionInstaller(SonarRuntime sqRuntime, DefaultPluginRepository pluginRepository) {
+    this.sqRuntime = sqRuntime;
     this.pluginRepository = pluginRepository;
   }
 
-  public ExtensionInstaller install(ComponentContainer container, ExtensionMatcher matcher) {
+  public ExtensionInstaller install(ComponentContainer container) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionInstaller.java]: MAJOR The Cyclomatic Complexity of this method "install" is 12 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionUtils.java [hunk 4]
```diff
@@ -37,8 +39,14 @@ public static boolean isInstantiationStrategy(Object extension, String strategy)
...
   }
 
-  public static boolean isBatchSide(Object extension) {
-    return AnnotationUtils.getAnnotation(extension, BatchSide.class) != null;
+  public static boolean isSonarLintSide(Object extension) {
+    return AnnotationUtils.getAnnotation(extension, SonarLintSide.class) != null;
+  }
+
+  public static boolean isScannerSide(Object extension) {
+    return AnnotationUtils.getAnnotation(extension, BatchSide.class) != null ||
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionUtils.java]: MINOR Remove this use of "BatchSide"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/standalone/rule/StandaloneActiveRulesProvider.java [hunk 5]
```diff
@@ -56,6 +69,22 @@ public ActiveRules provide(ProfileDefinition[] profileDefinitions) {
         registerProfilesForLanguage(builder, language, defs);
       }
 
+      for (Repository repo : ruleDefsLoader.getContext().repositories()) {
+        for (Rule rule : repo.rules()) {
+          if (rule.activatedByDefault()) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/standalone/rule/StandaloneActiveRulesProvider.java]: MAJOR Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)

File: core/src/main/java/org/sonarsource/sonarlint/core/util/StringUtils.java [hunk 6]
```diff
@@ -35,5 +35,20 @@ public static String urlEncode(String toEncode) {
...
+  
+  public static String describe(Object o) {
+    try {
+      if (o.getClass().getMethod("toString").getDeclaringClass() != Object.class) {
+        String str = o.toString();
+        if (str != null) {
+          return str;
+        }
+      }
+    } catch (Exception e) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/util/StringUtils.java]: MAJOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: core/src/main/java/org/sonarsource/sonarlint/core/analyzer/sensor/SensorsExecutor.java [hunk 7]
```diff
@@ -31,6 +31,7 @@
 import org.sonar.api.batch.events.SensorsPhaseHandler;
 import org.sonar.api.batch.events.SensorsPhaseHandler.SensorsPhaseEvent;
 import org.sonar.api.resources.Project;
+import org.sonarsource.sonarlint.core.util.StringUtils;
 
 @BatchSide
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/analyzer/sensor/SensorsExecutor.java]: MINOR Remove this use of "BatchSide"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues
* BLOCKER 1 blocker
* MAJOR 5 major
* MINOR 5 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [StandaloneActiveRulesProvider.java#L132](https://github.com/SonarSource/sonarlint-core/blob/19228e08e183ea28440ec390175fe1fdc1e4a555/core/src/main/java/org/sonarsource/sonarlint/core/container/standalone/rule/StandaloneActiveRulesProvider.java#L132): Make this anonymous inner class a lambda [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1604)
1. MINOR [ExtensionInstaller.java#L65](https://github.com/SonarSource/sonarlint-core/blob/19228e08e183ea28440ec390175fe1fdc1e4a555/core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionInstaller.java#L65): Remove this use of "ExtensionProvider"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. MINOR [ExtensionInstaller.java#L65](https://github.com/SonarSource/sonarlint-core/blob/19228e08e183ea28440ec390175fe1fdc1e4a555/core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionInstaller.java#L65): Remove this use of "ExtensionProvider"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. MINOR [ExtensionInstaller.java#L66](https://github.com/SonarSource/sonarlint-core/blob/19228e08e183ea28440ec390175fe1fdc1e4a555/core/src/main/java/org/sonarsource/sonarlint/core/container/global/ExtensionInstaller.java#L66): Remove this use of "ExtensionProvider"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Pull request merged**